library/gcm.c

```diff
@@ -337,23 +337,36 @@ int mbedtls_gcm_update_ad( mbedtls_gcm_context *ctx,
                            const unsigned char *add, size_t add_len )
 {
     const unsigned char *p;
-    size_t use_len, i;
+    size_t use_len, i, offset;
 
     GCM_VALIDATE_RET( add_len == 0 || add != NULL );
 
     /* IV is are limited to 2^64 bits, so 2^61 bytes */
     if( (uint64_t) add_len >> 61 != 0 )
         return( MBEDTLS_ERR_GCM_BAD_INPUT );
 
-    /* Calling update_ad multiple times is not yet supported */
-    if( ctx->add_len != 0 )
-        return( MBEDTLS_ERR_GCM_BAD_INPUT );
-
-    ctx->add_len = add_len;
+    offset = ctx->add_len % 16;
     p = add;
-    while( add_len > 0 )
+
+    if (offset)
+    {
+        use_len = 16 - offset;
```

Comment:
magic value 16 is used commonly here, probably a good idea to just use the length of `ctx->buf`

Reply:
The length of the buffer is not a good value here: it's the maximum supported block length, not the block length for the algorithm. If we ever support algorithms with a different length (which is not going to happen anytime soon), having `sizeof(buf)` would be worse than `16` since 16 is clearly a magic number but `sizeof(buf)` is correct when initializing and wiping buffers. So please keep 16 as a magic value as in the existing code.